typed-racket-lib/typed-racket/typecheck/tc-app/tc-app-objects.rkt

```diff
@@ -25,10 +25,21 @@
 (define-tc/app-syntax-class (tc/app-objects expected)
   #:literal-sets (kernel-literals object-literals)
   (pattern (dmo b cl
-            (#%plain-app list . pos-args)
-            (#%plain-app list (#%plain-app cons (quote names) named-args) ...))
+            (~and pos-arg-list (#%plain-app list . pos-args))
+            (~and named-arg-list (#%plain-app list (#%plain-app cons (quote names) named-args) ...)))
      #:declare dmo (id-from 'do-make-object 'racket/private/class-internal)
-     (check-do-make-object #'cl #'pos-args #'(names ...) #'(named-args ...)))
+     (begin0
+       (check-do-make-object #'cl #'pos-args #'(names ...) #'(named-args ...))
+       ;; synthesize a type for #'pos-arg-list, for the optimizer
+       (add-typeof-expr #'pos-arg-list
```

Comment:
`check-do-make-object` does not give a type to the list itself. It doesn't even see the list. That's why I have to reconstruct one here, otherwise that piece of code has no type.